prow/crier/reporters/gerrit/reporter.go

```diff
@@ -432,17 +456,58 @@ func GenerateReport(pjs []*v1.ProwJob, commentSizeLimit int) JobReport {
 		return true
 	})
 
-	// Truncate if there is not enough space left
 	remainingSize := commentSizeLimit - len(report.Header)
+	linesWithURLs := make([]string, len(report.Jobs))
+	linesWithoutURLs := make([]string, len(report.Jobs))
 	for i, job := range report.Jobs {
-		message := job.serialize() + "\n"
-		remainingSize -= len(message)
-		if remainingSize < 0 {
-			report.Message += fmt.Sprintf("[Skipped %d/%d jobs due to reaching gerrit comment size limit]\n", len(report.Jobs)-i, len(report.Jobs))
-			break
+		linesWithURLs[i] = job.serialize()
+		remainingSize -= len(linesWithURLs[i])
+	}
+
+	// cutoff is the index where if it's the line contains URL it will exceed the
+	// size limit.
+	cutoff := len(report.Jobs)
+	for cutoff > 0 && remainingSize < 0 {
+		cutoff--
+		linesWithoutURLs[cutoff] = report.Jobs[cutoff].serializeWithoutURL()
+		// remainingSize >= 0 after this condition means next line is not good
+		// to include URL
+		remainingSize += len(linesWithURLs[cutoff]) - len(linesWithoutURLs[cutoff])
+	}
+
+	// Should recommend /test all when URL is skipped on failed/aborted prowjobs
+	if cutoff != len(report.Jobs) &&
+		(report.Jobs[cutoff].State == v1.FailureState || report.Jobs[cutoff].State == v1.AbortedState || report.Jobs[cutoff].State == v1.ErrorState) {
+		reTestMessage = " Comment '/test all' to rerun all failed tests"
+	}
+
+	// Now there is enough info, we can finalize Header
+	// Note that the change in reTestMessage makes the comment longer, but since
+	// the size limit of 14400 is conservative, we should be fine
+	report.Header += fmt.Sprintf("%s\n", reTestMessage)
+
+	// This shouldn't happen unless there are too many prow jobs(e.g. > 300) and
+	// each job name is super long(e.g. > 50)
+	if remainingSize < 0 {
+		report.Message = "Prow failed to report all jobs, are there excessive amount of prow jobs?"
```

Comment:
After a second thought, I think the only case where we should recommend `/test all` is here. Because after this change `/retest` should work as expected.

Reply:
done, @cjwagner please take another look